lib/logstash/outputs/elasticsearch/http_client/pool.rb

```diff
@@ -282,19 +284,10 @@ def healthcheck!
               @logger.warn("Detected a node with a higher major version than previously observed. This could be the result of an elasticsearch cluster upgrade.", :previous_major => @maximum_seen_major_version, :new_major => major, :node_url => url.sanitized.to_s)
               set_new_major_version(major)
             end
-            if oss? || valid_es_license?(url)
-              meta[:state] = :alive
-            else
-              # As this version is to be shipped with Logstash 7.x we won't mark the connection as unlicensed
-              #
-              #  logger.error("Cannot connect to the Elasticsearch cluster configured in the Elasticsearch output. Logstash requires the default distribution of Elasticsearch. Please update to the default distribution of Elasticsearch for full access to all free features, or switch to the OSS distribution of Logstash.", :url => url.sanitized.to_s)
-              #  meta[:state] = :unlicensed
-              #
-              # Instead we'll log a deprecation warning and mark it as alive:
-              #
-              log_license_deprecation_warn(url)
-              meta[:state] = :alive
-            end
+
+            # defined in license_check.rb
+            license_check!(url, meta)
```

Comment:
Maybe a pre-flight proc (`connection_preflight`?) could be registered by the thing creating the client pool, which would enable different common consumers to do their own pre-flights, perhaps including their own specific license-checking behaviour or registering a no-op preflight if there is no extra steps needed.

Reply:
🤔

Reply:
The license checking logic is badly intertwined in the Pool logic and I am a bit worried about performing this refactor at this stage. I agree that it would be a positive code cleanup but I am unsure it is worth the investment now-now just for the benefit of making the data_streams output work. This proposed solution is rather non-intrusive and I will add more comments about the intent. Maybe keep a tab on this as a future refactoring followup?

Reply:
you can disregard this above comment, I moved that discussion at the top level.